haystack/preview/components/classifiers/file_classifier.py

```diff
@@ -0,0 +1,75 @@
+import logging
+from collections import defaultdict
+from pathlib import Path
+from typing import List, Union, Dict
+
+from haystack.preview import component
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class FileTypeClassifier:
+    """
+    A component that classifies files based on their file extensions.
+
+    The FileTypeClassifier takes a list of file paths and groups them by their file extensions.
+    The list of extensions to consider is provided during the initialization of the component.
+
+    This component is particularly useful when working with a large number of files, and you
+    want to categorize them based on their file types.
+    """
+
+    @component.input
+    def input(self):
+        """
+        The input data for the FileTypeClassifier.
+
+        It expects a list of file paths. The file paths can be of type str or pathlib.Path.
+        """
+
+        class Input:
+            paths: List[Union[str, Path]]
+
+        return Input
+
+    @component.output
+    def output(self):
+        """
+        The output data of the FileTypeClassifier.
+
+        It returns a dictionary where the keys are the file extensions, and the values are lists
+        of pathlib.Path objects, representing the file paths that match the corresponding extension.
+        """
+
+        class Output:
+            extensions: Dict[str, List[Path]]
+
+        return Output
```

Comment:
Although this works, I believe we can do better :slightly_smiling_face: 

Right now you're making a component with one output. That means that all components that all components that connect to this one will receive the whole dictionary, and they have to know which key contains the documents they need.

However file converters do not accept a dictionary: they expect the list of files and nothing more. See `TextFileToDocument`: https://github.com/deepset-ai/haystack/blob/f6b50cfdf96daa4f84f2d4e93edd8eb1e11e0b89/haystack/preview/components/file_converters/txt.py#L57

So how can we connect this to the file converters?

The idea is that `FileTypeClassifier` should have several outputs, one for each file type. It's not obvious how to do it, but it's possible, because `def output(self)` has all the information it needs: what you have to do is to build the output dataclass dynamically.

Assuming that in the init method you're saving the list of extensions as `self.extensions`, here is how you could do it:

```python
from dataclasses import make_dataclass

...

@component.output
    def output(self):
        """
        The output data of the FileTypeClassifier.
        it returns as many outputs as self.extensions keys.
        """
        return make_dataclass(
            "Output", fields=[(extension, Optional[List[Union[str, Path]]], None) for extension in self.extensions]
        )
```

Of course then you'll need to adjust many other parts of the code, like the output of run(), the tests etc. I'll not comment on those to avoid spamming this PR.